lib/collection/src/shards/local_shard/clock_map.rs

```diff
@@ -0,0 +1,121 @@
+use std::collections::HashMap;
+use std::io::Write as _;
+use std::path::Path;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::{fs, io};
+
+use serde::{Deserialize, Serialize};
+
+use crate::operations::types::CollectionError;
+use crate::operations::ClockTag;
+use crate::shards::shard::PeerId;
+
+#[derive(Debug, Default, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct ClockMap {
+    clocks: HashMap<ClockId, Clock>,
+}
+
+impl ClockMap {
+    pub fn load_or_default(path: &Path) -> Result<Self> {
+        let result = Self::load(path);
+
+        if let Err(Error::Io(err)) = &result {
+            if err.kind() == io::ErrorKind::NotFound {
+                return Ok(Self::default());
+            }
+        }
+
+        result
+    }
+
+    pub fn load(path: &Path) -> Result<Self> {
+        let file = fs::File::open(path)?;
+        let clock_map = serde_json::from_reader(io::BufReader::new(file))?;
+        Ok(clock_map)
+    }
+
+    pub fn store(&self, path: &Path) -> Result<()> {
+        let file = fs::File::create(path)?;
+
+        let mut writer = io::BufWriter::new(file);
+        serde_json::to_writer(&mut writer, &self)?;
+        writer.flush()?;
+
+        Ok(())
+    }
+
+    pub fn advance_clock_and_correct_tag(&mut self, clock_tag: &mut ClockTag) -> u64 {
+        let prev_tick = self.advance_clock(clock_tag);
+
+        if clock_tag.clock_tick == 0 {
+            clock_tag.clock_tick = prev_tick;
+        }
+
+        prev_tick
+    }
+
+    pub fn advance_clock(&mut self, clock_tag: &ClockTag) -> u64 {
+        let clock_id = ClockId::from_tag(clock_tag);
+        let new_tick = clock_tag.clock_tick;
+
+        if let Some(clock) = self.clocks.get(&clock_id) {
+            clock.advance_to(new_tick)
+        } else {
+            self.clocks.insert(clock_id, Clock::new(new_tick));
+            0
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, Deserialize, Serialize)]
+struct ClockId {
+    peer_id: PeerId,
+    clock_id: u32,
+}
+
+impl ClockId {
+    pub fn from_tag(clock_tag: &ClockTag) -> Self {
+        Self {
+            peer_id: clock_tag.peer_id,
+            clock_id: clock_tag.clock_id,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+struct Clock {
+    clock: AtomicU64,
+}
+
+impl Clock {
+    pub fn new(tick: u64) -> Self {
+        Self {
+            clock: AtomicU64::new(tick),
+        }
+    }
+
+    pub fn advance_to(&self, new_tick: u64) -> u64 {
+        self.clock.fetch_max(new_tick, Ordering::Relaxed)
+    }
+}
```

Comment:
Can't we reuse the same `Clock` type here, to prevent name collisions?

Reply:
We can't reuse them, cause they work differently. Like, `advance_to` is implemented differently on `clock_set::Clock` and `clock_map::Clock`.

Also, these are private types, so I don't think any kind of name collision is possible. %)

Reply:
Is there no way to unify them?

> Also, these are private types, so I don't think any kind of name collision is possible. %)

I meant that we have the same name for a similar-but-not-the-same type, which is very confusing to me.

Reply:
> Is there no way to unify them?
Maybe there is, but I don't want to, tbh.

> similar-but-not-the-same type
Yeah, exactly. In my mind they are, logically, two different types. _Somewhat_ similar, sure, but even the main method (i.e., `advance_to`) is different for both of them.

And regarding type names... I dunno. I personally prefer this approach of using simple, clear names (even if the same name is used for multiple types) and just scope them with modules.

Like, I actually *don't* like when we name our types `CollectionError` or `CollectionResult`. IMO using `collection::Error`/`collection::Result` is just better. It's the same amount of typing (or, two additional symbols for `::`), but *inside* the `collection` module you just use `Error` instead of repeating `Collection*` all the time and this way you import less stuff (cause you just do `use collection` and then do `collection::Error` instead of `use collection::CollectionError` and then also `CollectionError`).

Does this make sense?

Reply:
I like short names to. But with this approach you always need to know or define your current context, and you need to know what file/module your cursor is currently in. That has always been a problem for me personally, requiring me to jump/look at things a lot more when working with types like this. With unique names I don't have to think about this at all.

I have a similar problem with all the Mutex/RwLock types we're using interchangeably, to me it is extremely annoying.

But whatever, lets see if someone else has an opinion on this.

Reply:
> But with this approach you always need to know or define your current context, and you need to know what file/module your cursor is currently in.

Is it a problem even in case of `ClockSet`/`ClockMap`? Like, it's a private type used only as part of `ClockSet`/`ClockMap` implementation and both these modules are 100 something lines of code.

I'm not completely against renaming these to, e.g., `ClockSetClock`/`ClockMapClock`... but, tbh, it feels so "extra", given these are private types used in a very limited context. Just rubs me the wrong way. But I'm ok renaming them, if it makes it easier for everyone else to work with. 😭

> I have a similar problem with all the Mutex/RwLock types...

Same. I hate this.

IMO, typing module name explicitly (e.g., `parking_lot::Mutex` and `tokio::sync::Mutex`) solves the issue in the best way.

Reply:
We agreed to keep it like this for now. If I come up with a good example on how we might unify this, we can make the necessary changes in the future.